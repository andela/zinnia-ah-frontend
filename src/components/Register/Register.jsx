import React, { Component } from 'react';
import swal from 'sweetalert';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

// helper functions
import { signupUser } from '../../store/modules/auth';
import {
  validateInput,
  checkAllEmptyFields,
  fieldChecker,
} from '../../utils/formValidator';

// components
import Button from '../presentationals/Button/Button';
import Image from '../presentationals/Image/Image';
import Loader from '../presentationals/Loader/Loader';

// styles
import '../../components/presentationals/AuthenticationCard/AuthenticationCard.scss';
import '../../components/presentationals/Form.scss';

// images
import GoogleIcon from '../../assets/images/google-icon.svg';
import FacebookIcon from '../../assets/images/facebook-icon.svg';
import TwitterIcon from '../../assets/images/twitter-icon.svg';

export class Register extends Component {
  state = {
    userCredentials: {
      email: '',
      password: '',
      username: '',
    },
    validationErrors: [],
    emptyFields: [],
  };

  userInputHandler = event => {
    const { emptyFields } = this.state;
    const currentEmptyFields = fieldChecker(event, emptyFields);
    this.setState({
      userCredentials: {
        ...this.state.userCredentials,
        [event.target.name]: event.target.value,
      },
      emptyFields: currentEmptyFields,
    });
  };

  formValidator = event => {
    event.preventDefault();
    const { userCredentials } = this.state;
    const emptyFields = checkAllEmptyFields(userCredentials);
    if (emptyFields.length) {
      this.setState({
        emptyFields,
      });
      return false;
    }
    this.setState({ emptyFields: [] });
    const validationErrors = validateInput(userCredentials);
    this.setState({
      validationErrors,
    });
    if (!validationErrors.length) {
      this.setState({ validationErrors: '' });
      return true;
    }
    return false;
  };

  submitForm = event => {
    event.preventDefault();
    this.props.signupUser({
      ...this.state.userCredentials,
    });
    const userCredentials = { email: '', password: '', username: '' };
    this.setState({
      userCredentials,
    });
  };

  render() {
    const { isLoading, errorResponse, successResponse } = this.props.auth;
    const { emptyFields } = this.state;
    successResponse.status === 'success' &&
      swal('congratulations', successResponse.message, 'success');
    return (
      <div>
        {isLoading && <Loader text="please wait" size="large" />}
        <form
          id="form"
          className="form"
          onSubmit={e => {
            this.formValidator(e) && this.submitForm(e);
          }}
        >
          <div className="form-group">
            <input
              className={`form-control ${
                emptyFields.includes('username') ? 'error' : ''
              }`}
              placeholder="Username"
              onChange={this.userInputHandler}
              value={this.state.userCredentials.username}
              name="username"
            />
          </div>

          <div className="form-group">
            <input
              placeholder="Email"
              className={`form-control ${
                emptyFields.includes('email') ? 'error' : ''
              }`}
              type="email"
              onChange={this.userInputHandler}
              value={this.state.userCredentials.email}
              name="email"
            />
          </div>
          <div className="form-group">
            <input
              className={`form-control ${
                emptyFields.includes('password') ? 'error' : ''
              }`}
              placeholder="Password"
              type="password"
              onChange={this.userInputHandler}
              value={this.state.userCredentials.password}
              name="password"
            />
          </div>
          {this.state.validationErrors.length > 0 && (
            <div className="ui negative message">
              {this.state.validationErrors.map(error => (
                <p key={error}>{error}</p>
              ))}
            </div>
          )}
          {errorResponse.length > 0 && (
            <div className="ui negative message">
              {errorResponse.map(error => (
                <p key={error.message}>{error.message}</p>
              ))}
            </div>
          )}
          <Button type="submit" value="GET STARTED" className="btn-dark" />
        </form>
        <div className="d-flex or-div">
          <hr className="hr" />
          <p className="or">OR</p>
          <hr className="hr" />
        </div>
        <div className="icon-container">
          <div className="icon-card">
            <Image src={GoogleIcon} />
          </div>
          <div className="icon-card">
            <Image src={FacebookIcon} />
          </div>
          <div className="icon-card">
            <Image src={TwitterIcon} />
          </div>
        </div>
      </div>
    );
  }
}
Register.propTypes = {
  signupUser: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
};

const mapStateToProps = state => ({ auth: state.auth });

export default connect(
  mapStateToProps,
  { signupUser },
)(Register);
