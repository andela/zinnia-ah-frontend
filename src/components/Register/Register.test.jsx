import React from 'react';
import { mount } from 'enzyme';

import { Register } from '../Register/Register';

const props = {
  auth: {
    isLoading: true,
    successResponse: {
      status: 'success',
      message: '',
    },
    errorResponse: [],
  },
  signupUser: jest.fn(),
};

const mountWrapper = mount(<Register {...props} />);

describe('<Register />', () => {
  it('mounts correctly', () => {
    expect(mountWrapper).toMatchSnapshot();
  });

  it('renders a form with className form', () => {
    const form = mountWrapper.find('form');
    expect(form.hasClass('form')).toBe(true);
  });

  it('rendered a form with 3 input fields', () => {
    const inputs = mountWrapper.find('input').getElements();
    expect(inputs.length).toBe(3);
  });

  it('correctly initializes the value of the username field to an empty string', () => {
    const usernameField = mountWrapper.find('input[name="username"]');
    expect(usernameField.prop('value')).toBe('');
    const emailField = mountWrapper.find('input[name="email"]');
    expect(emailField.prop('value')).toBe('');
    const passwordField = mountWrapper.find('input[name="password"]');
    expect(passwordField.prop('value')).toBe('');
  });

  it('does not submit the form if there is an empty field', () => {
    const form = mountWrapper.find('form');
    form.simulate('submit', { preventDefault: jest.fn() });
    expect(mountWrapper.instance().state.emptyFields.length).toBe(3);
    expect(mountWrapper.instance().state.emptyFields).toContain(
      'email',
      'password',
      'username',
    );
    const input = mountWrapper.find('input[name="username"]');
    expect(input.props().className).toContain('error');
  });

  it('updates all fields with invalid values', () => {
    const username = mountWrapper.find('input[name="username"]');
    username.instance().value = 'io';
    username.simulate('change');
    const email = mountWrapper.find('input[name="email"]');
    email.instance().value = 'igbominadeveloper@';
    email.simulate('change');
    const password = mountWrapper.find('input[name="password"]');
    password.instance().value = 'password1____';
    password.simulate('change');
    expect(mountWrapper.state().userCredentials.username).toBe('io');
    expect(mountWrapper.state().userCredentials.email).toBe(
      'igbominadeveloper@',
    );
    expect(mountWrapper.state().userCredentials.password).toBe('password1____');
    const form = mountWrapper.find('form');
    form.simulate('submit', { preventDefault: jest.fn() });
    expect(mountWrapper.state().validationErrors.length).toBeGreaterThan(0);
  });

  it('returns a validation error for an invalid field input', () => {
    const username = mountWrapper.find('input[name="username"]');
    username.instance().value = 'igbominadeveloper';
    username.simulate('change');
    const email = mountWrapper.find('input[name="email"]');
    email.instance().value = 'igbominadeveloper@ah.com';
    email.simulate('change');
    const password = mountWrapper.find('input[name="password"]');
    password.instance().value = 'passwor';
    password.simulate('change');
    expect(mountWrapper.state().userCredentials.username).toBe(
      'igbominadeveloper',
    );
    expect(mountWrapper.state().userCredentials.email).toBe(
      'igbominadeveloper@ah.com',
    );
    expect(mountWrapper.state().userCredentials.password).toBe('passwor');
    const form = mountWrapper.find('form');
    form.simulate('submit', { preventDefault: jest.fn() });
    expect(mountWrapper.state().validationErrors.length).toBeGreaterThan(0);
    password.instance().value = '';
    password.simulate('change');
    expect(mountWrapper.state().validationErrors.length).toBeGreaterThan(0);
  });

  it('should submit the form and have an empty error body', () => {
    const username = mountWrapper.find('input[name="username"]');
    username.instance().value = 'igbominadeveloper';
    username.simulate('change');
    const email = mountWrapper.find('input[name="email"]');
    email.instance().value = 'igbominadeveloper@ah.com';
    email.simulate('change');
    const password = mountWrapper.find('input[name="password"]');
    password.instance().value = 'password1';
    password.simulate('change');
    const form = mountWrapper.find('form');
    form.simulate('submit', { preventDefault: jest.fn() });
    expect(mountWrapper.state().validationErrors.length).toBe(0);
  });
});
