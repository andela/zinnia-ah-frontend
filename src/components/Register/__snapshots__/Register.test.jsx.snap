// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Register /> mounts correctly 1`] = `
<Register
  auth={
    Object {
      "errorResponse": Array [],
      "isLoading": true,
      "successResponse": Object {
        "message": "",
        "status": "success",
      },
    }
  }
  signupUser={[MockFunction]}
>
  <div>
    <Loader
      size="large"
      text="please wait"
    >
      <div
        className="ui active inverted dimmer"
      >
        <div
          className="ui indeterminate text loader large"
        >
          please wait
        </div>
      </div>
    </Loader>
    <form
      className="form"
      id="form"
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <input
          className="form-control "
          name="username"
          onChange={[Function]}
          placeholder="Username"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="form-control "
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="form-control "
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
      <Button
        className="btn-dark"
        type="submit"
        value="GET STARTED"
      >
        <button
          className="btn btn-dark"
          type="submit"
        >
          GET STARTED
        </button>
      </Button>
    </form>
    <div
      className="d-flex or-div"
    >
      <hr
        className="hr"
      />
      <p
        className="or"
      >
        OR
      </p>
      <hr
        className="hr"
      />
    </div>
    <div
      className="icon-container"
    >
      <div
        className="icon-card"
      >
        <Image
          src=""
        >
          <img
            src=""
          />
        </Image>
      </div>
      <div
        className="icon-card"
      >
        <Image
          src=""
        >
          <img
            src=""
          />
        </Image>
      </div>
      <div
        className="icon-card"
      >
        <Image
          src=""
        >
          <img
            src=""
          />
        </Image>
      </div>
    </div>
  </div>
</Register>
`;
